File: tam_agent/research/data_extraction.py
"""Extract structured market data from search results using LLM."""

from __future__ import annotations

from typing import TYPE_CHECKING

from pydantic import BaseModel, Field

from tam_agent.models import MarketDataPoint, SearchResult

if TYPE_CHECKING:
    from tam_agent.llm.client import LLMClient


class ExtractedData(BaseModel):
    """Container for extracted data points."""
    data_points: list[MarketDataPoint] = Field(default_factory=list)


def extract_market_data(
    llm: LLMClient,
    search_results: list[SearchResult],
    context: str,
) -> list[MarketDataPoint]:
    """Extract structured market data from search result snippets."""
    if not search_results:
        return []

    snippets_text = "\n\n".join(
        f"Source: {r.title} ({r.url})\n{r.snippet}"
        for r in search_results
    )

    prompt = (
        f"Extract quantitative market data points from the following search results.\n\n"
        f"Context: {context}\n\n"
        f"Search Results:\n{snippets_text}\n\n"
        f"For each data point, identify:\n"
        f"- The metric being measured (e.g., 'global cybersecurity market size')\n"
        f"- The value (e.g., '$180 billion')\n"
        f"- The reference year if mentioned\n"
        f"- The source URL and name\n"
        f"- Reliability: 'official_report' for industry reports/government data, "
        f"'news_article' for news/press, 'estimate' for analyst/blog estimates\n\n"
        f"Only extract data points with concrete numerical values. If no quantitative "
        f"data is found, return an empty list."
    )

    result = llm.structured_query(
        prompt=prompt,
        response_model=ExtractedData,
        system="You are a data extraction specialist. Extract only factual numerical data.",
    )
    return result.data_points


def format_search_data_section(
    search_results: list[SearchResult],
    data_points: list[MarketDataPoint],
) -> str:
    """Format search data into a text section for prompt injection."""
    if not search_results and not data_points:
        return (
            "Note: No web search data was available. Base your estimates on your "
            "training knowledge, and rate confidence lower accordingly."
        )

    parts: list[str] = []

    if data_points:
        parts.append("Extracted Market Data Points:")
        for dp in data_points:
            year_str = f" ({dp.year})" if dp.year else ""
            parts.append(f"  - {dp.metric}: {dp.value}{year_str} [Source: {dp.source_name}]")

    if search_results:
        parts.append("\nRaw Search Snippets:")
        for r in search_results[:10]:
            parts.append(f"  - {r.title}: {r.snippet[:200]}")

    return "\n".join(parts)

File: src/research/data-extraction.ts
/** Extract structured market data from search results using LLM. */

import { z } from "zod";
import type { LLMClient } from "../llm/client.js";
import { MarketDataPointSchema, type MarketDataPoint, type SearchResult } from "../models.js";

const ExtractedDataSchema = z.object({
  data_points: z.array(MarketDataPointSchema).default([]),
});

export async function extractMarketData(
  llm: LLMClient,
  searchResults: SearchResult[],
  context: string,
): Promise<MarketDataPoint[]> {
  if (searchResults.length === 0) return [];

  const snippetsText = searchResults
    .map((r) => `Source: ${r.title} (${r.url})\n${r.snippet}`)
    .join("\n\n");

  const prompt =
    `Extract quantitative market data points from the following search results.\n\n` +
    `Context: ${context}\n\n` +
    `Search Results:\n${snippetsText}\n\n` +
    `For each data point, identify:\n` +
    `- The metric being measured (e.g., 'global cybersecurity market size')\n` +
    `- The value (e.g., '$180 billion')\n` +
    `- The reference year if mentioned\n` +
    `- The source URL and name\n` +
    `- Reliability: 'official_report' for industry reports/government data, ` +
    `'news_article' for news/press, 'estimate' for analyst/blog estimates\n\n` +
    `Only extract data points with concrete numerical values. If no quantitative ` +
    `data is found, return an empty list.`;

  const result = await llm.structuredQuery(
    prompt,
    ExtractedDataSchema,
    "ExtractedData",
    "You are a data extraction specialist. Extract only factual numerical data.",
  );
  return result.data_points ?? [];
}

export function formatSearchDataSection(
  searchResults: SearchResult[],
  dataPoints: MarketDataPoint[],
): string {
  if (searchResults.length === 0 && dataPoints.length === 0) {
    return (
      "Note: No web search data was available. Base your estimates on your " +
      "training knowledge, and rate confidence lower accordingly."
    );
  }

  const parts: string[] = [];

  if (dataPoints.length > 0) {
    parts.push("Extracted Market Data Points:");
    for (const dp of dataPoints) {
      const yearStr = dp.year ? ` (${dp.year})` : "";
      parts.push(
        `  - ${dp.metric}: ${dp.value}${yearStr} [Source: ${dp.source_name}]`,
      );
    }
  }

  if (searchResults.length > 0) {
    parts.push("\nRaw Search Snippets:");
    for (const r of searchResults.slice(0, 10)) {
      parts.push(`  - ${r.title}: ${r.snippet.slice(0, 200)}`);
    }
  }

  return parts.join("\n");
}
